### ISSUE TITLE
Importer sometimes skips duplicates without prompting for a choice

### ISSUE THREAD

**Issue by ekjaker** [sampsyo/beets#1652]
Beets will skip sometimes duplicate albums when a prompt is needed.  I found this to happen quite often with large imports, especially after after the 20-30 mark where more than 50% gets trumped unjustly. The problem is rather annoying and potentially dangerous, as it creates empty folders without bringing in the audiofiles. The only way to know that something went wrong is check the log file and correct manually, which due to the high occurrence is not realistically feasible.

For every release, I always have a mp3 and flac folder. When importing, it should detect the duplicate and then ask whether I want to keep, release or remove, which it often does. But not always. The result is that I get both a folder for the flac and the mp3 version (I have paths setup to separate them), but one of them won't have any audiofiles, although the .cue .log file might be there from copyartifacts. The output on screen is "This album is already in the library! All duplicates are empty, we ignore them". In the logfile, I get a "duplicate-replace". This only happens with duplicates that get accepted automatically (t my knowledge). When I to have manually apply, it will give the duplicate prompt.

I did some basic troubleshooting. When checking the database afterwards with ls, I only get one of the releases, either the flac or the mp3. When manually adding just the missing one, the duplicate question pops up and it gets added without an issue, showing every file twice with ls. When removing the database entry for that release (both mp3 files and flac files), and re-importing both folders manually again, again there is no problem.
I also tried it without being threaded: no issue.

My guess is that when working with a large import copying the files to a new location (which is what I'm doing), a large queue of files to be copied might stack up, especially when there are a lot of highly recommended releases that get processed faster than they can get copied. When such a copy-backlog is present, and 2 duplicate folders (with different formats) come immediately one after the other, the first one gets processed normally and the copy operation is added to the copy-queue. When the second one gets processed, beets detects a duplicate in the database, checks the folder but finds it still to be empty (because the copy operation is still in the backlog) and does a duplicate-replace.

To reproduce the bug is not going be very straight forward, I'm afraid. I suggest preparing a folder with plenty of releases that get recognized easily and automatically get tagged as highly recommended (about 20 full size albums should be sufficient), followed by a few releases for duplicate albums but different formats. Probably the easiest way to do it, is to a run an import on a basic folder with a lot of releases till you have done about 40. Remove all the ones that needed input. Add the duplicate albums/different formats. If necessary, rename the folders of these albums to make them start with a "z" or something at the back of the alphabet, as those will be processed last by the importer. I have the error when performing an import-copy from one HDD to another. SSD might be too fast to easily reproduce the error.

**Comment by sampsyo** [comment]
To quote myself on the mailing list:

> I see two problems here:
> - Albums should *never* be added empty. The insertion of the album structure should happen atomically, in a transaction that prevents any other threads from seeing an empty album. The transaction handling is pretty careful about this, but somehow that's not working.
> - There is a race from when an album is approved by the user and when it is available for conflict detection. (The UI happens in the `user_query` thread; the album is actually added in `apply_choices`.) I don't remember when we split this up, but it was a mistake and we need to revert it.

Do you think you could try putting together a bundle of music that (at least sometimes) triggers the problem for you? That would be an amazing starting point for getting to the bottom of the problem.

**Comment by ekjaker** [comment]
I think I could, but I'm afraid it would not suitable for distribution. PM?

**Comment by sampsyo** [comment]
Yes, good thinking! My email's on my GitHub profile page.

**Comment by ekjaker** [comment]
I wanted to create a folder with as little releases as possible in it, but that was still able to recreate the error so I could forward that. So I've copied the folders which caused the issue to another computer (for easier editing) and gave it another go so I could trim out the releases that needed input for quicker testing: this time all went through without error.:-(
I have no idea what might be causing this difference, except maybe for the fact that the disk on the second pc is faster, although it did manage to gather up a high copy-backlog as well (after asking the last release, it still had 10-15 minutes of copying left). First is Windows 7, the second Windows 10. First is a 2-disk operation, the second a 1 disk. 
Went back to the first pc, for a new test after deleting the library (to make sure there was nothing there that might be causing the issue): same issue again. I have to add, on this run the issue started  only after processing about 15GB of releases (about as much mp3 as FLAC), but when it started, it was very frequent.
I'll be copying everything to pc 1 and doing my tagging from there, keeping my eye on the logfile for duplicate-replace.
If you have any suggestions or requests, please let me know. I could create a bundle of music as requested, but it'll be huge to upload and might not reproduce the same issue on every pc, so probably not the best way to go.

**Comment by sampsyo** [comment]
Huh, that's certainly frustrating. Thanks for looking into it!

If you can even get this to reproduce consistently on one machine, that could be helpful by itself. If, for example, you can get a verbose log from a run that exhibits the error, we can start there. I'll try applying a fix and ask you to run it again to see if we've actually fixed it.

**Comment by ekjaker** [comment]
Good news. After numerous lengthy test runs and a lucky observation, I was able to compile a bundle of music and a tagging procedure to consistently reproduce the error on both pc's, narrowing down the circumstances in which it occurs to some quite specific parameters. Instead of being caused by a copy queue, it seems to be triggered by a queue of releases that need input.

Basically the releases I'm testing in 1 import batch with autotagging, are all pairs of mp3 versus flac duplicate albums that get recognized as highly recommended. When running the import, it hits the first and second release and detects a duplicate. Instead of immediately responding, keep it waiting for a while so that other threads start resolving the other albums as well, creating a queue of releases that need input. When after a while, you finally do select to keep the duplicate, it scrolls really fast through the inputqueue to the last question in the queue, skipping all the ones in between with a skip-replace and the "All duplicates are empty, we ignore them" remark in the output. By doing so, it creates a folder for the duplicate with no audio files but with the artifacts in it if the plugin is activated. If the copyartifacts plugin is not activated, no extra folder is created. But in both instances, beets is skipping the duplicate I want to (have an option to) keep.

Running in verbose only creates a folder for the first hit, then displays an error "state file could not be read: [Errno 2] No such file or directory: u'C:\\Users\\couli\\AppData\\Roaming\\beets\\state.pickle'" , then the rolls through to the last release without errors and then hangs, requiring a KeyboardInterrupt.

FYI:
I tried it on both my personal config as the default: same issue.
I had beets create a new database on my test runs to eliminate possible issues with the database. I also deleted the state.pickle at every run.

I'll send you the music files, together with my log files and config files, so you can have a look yourself. And thank you in advance for doing so! To be honest, with this bug, I can't really use beets. I just have too many duplicates. And although I'm still very new to beets, I did really learn to appreciate all it can do and represents.

Let me know it if there is anything I can do to help.

Ekjaker

**Comment by sampsyo** [comment]
Perfect! Thank you so much for doing all this investigation; I'll see what I can find using your replication case.

**Comment by sampsyo** [comment]
Thank you again for getting together a replication case! This worked perfectly.

I tracked down the bug. It was an unhappy interaction between two pieces of code that I didn't like much anyway:
1. An explicit way of keeping track of recently-imported albums for the purpose of duplicate detection *without adding them to the library*. This was always fragile and, fortunately, unnecessary with the current importer design.
2. The UI code that decided to skip asking about duplicates when they were all empty. This was erroneously classifying recent conflicts as the all-empty case. (For the record, my original hypothesis was wrong: we were not, in fact, adding empty albums to the database.)

Removing either of these would have fixed the problem, but I dove in and removed them both. :smiley: This should fix the race you were seeing.

**Comment by ekjaker** [comment]
Fantastic! I found number 2 as well and was already thinking of going all dirty on the code and just commenting that part out, but this obviously beats that. 
Maybe a stupid question: if I install the beets source code instead of the regular 1.3.5 release, do I also get the new commits? Or is it better to just wait for the new 1.3.6 release? Actually, never mind. It is a stupid question as I can answer it myself by just giving it a try. :-) Should be doable with all the info in the wiki. And maybe in the future I just might try some small coding changes myself if I can find the time to invest in Python.
Thank you again for fixing this, and even more, doing it so fast!!
